### PR TITLE
[precompile] Add the inductor backend to torch.compiler.precompile

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #190226
* #189622
* #189165
* __->__ #189621

Builds on the eager backend from the previous PR in this stack. Adds
backend="inductor" (now the default), which lowers the captured make_fx graph
through torch._functorch.aot_autograd.compile_to_python (AOTAutograd + Inductor,
composed into one self-contained module). python_code inlines the Inductor output
(kernels JIT-compile on first call) plus AOTAutograd's codegen'd prelude/epilogue
(subclass wrap/unwrap, input-mutation reflection, output aliasing); the cache is
the save_cache_artifacts bundle that load() feeds to load_cache_artifacts to prime
the inductor kernel caches, so a warm reload loads precompiled kernels instead of
recompiling. With no usable cache, load() degrades to JIT'ing from python_code.

The inductor backend additionally specializes on each read input's stride / memory
format: size_asserts is pinned on so the artifact bakes assert_size_stride for
every input the graph reads (invariant 6), and the inlined driver relabels an
assert_size_stride failure as a clear PrecompileError.

Shapes remain static (dynamic shapes follow in the next PR).

Documentation for the precompile API (torch.compiler_api.md and the torch.compiler user guide) is split into a separate documentation PR at the top of this stack to keep this PR under the line limit; this PR is code only.

Test Plan:

```
python test/test_precompile.py
```

Authored with an AI assistant.

### CODE REVIEW

File: torch/_precompile.py [hunk 1]
```diff
@@ -956,14 +1079,204 @@ def forward(*args):
...
+            "match a contiguous example, or match the example weight's layout), or use "
+            "backend='eager'. Underlying: %s" % str(_e)
+        )
+    if GRAD_PARAM_INDICES:  # noqa: F821
+        n = len(GRAD_PARAM_INDICES)  # noqa: F821
+        grads = out[len(out) - n:]
+        out = out[:len(out) - n]
+        for idx, g in zip(GRAD_PARAM_INDICES, grads):  # noqa: F821
+            p = pb[idx]
+            p.grad = g if p.grad is None else p.grad + g
```
Comment (aorenste) [inline torch/_precompile.py]: [blocker] The inductor driver accumulates grads out-of-place, rebinding `.grad` to a new tensor; the eager driver (`_EAGER_DRIVER_SOURCE`) and this module's documented invariant 5 (`p.grad += g`) both accumulate in-place. Match them:

```suggestion
            if p.grad is None:
                p.grad = g
            else:
                p.grad.add_(g)
```

Out-of-place `p.grad + g` replaces `.grad` with a fresh object on every accumulating call, so a `zero_grad(set_to_none=False)` loop (or any caller/optimizer keyed on grad-tensor identity) behaves differently on the inductor backend than on the eager backend or eager `.backward()` -- the grad identity the capture path deliberately snapshots/restores.

File: torch/_precompile.py [hunk 2]
```diff
@@ -956,14 +1079,204 @@ def forward(*args):
...
+        )
+    mods = []
+    for _i in MODULE_POSITIONS:  # noqa: F821
+        if not isinstance(args[_i], _torch.nn.Module):
+            _fail(
+                "precompile: argument at position %d must be the nn.Module the traced "
+                "fn took (invariant 2), got %s." % (_i, type(args[_i]).__name__)
+            )
+        mods.append(args[_i])
+    user_inputs = [a for i, a in enumerate(args) if i not in set(MODULE_POSITIONS)]  # noqa: F821
```
Comment (aorenste) [inline torch/_precompile.py]: [cleanup] `set(MODULE_POSITIONS)` is rebuilt on every call here; the eager driver hoists it once to a module-level `_MODULE_POSITIONS_SET` (see `_EAGER_DRIVER_SOURCE`). Mirror that in `_DRIVER_SOURCE`: add `_MODULE_POSITIONS_SET = set(MODULE_POSITIONS)` at the top and use `i not in _MODULE_POSITIONS_SET` here.

File: torch/_precompile.py [hunk 3]
```diff
@@ -960,14 +1083,204 @@ def forward(*args):
 '''
 
 
+_DRIVER_SOURCE = '''
```
Comment (aorenste) [inline torch/_precompile.py:1086]: Same comment as in [#189620](https://github.com/pytorch/pytorch/pull/189620#discussion_r3597481361): Consider exporting this out to a separate module so it can be checked with tooling.

**Review by aorenste — DISMISSED** [review]
(Reviewed by me, assisted by AI)

[cleanup] The artifact-stripping idiom -- `blob = torch.load(io.BytesIO(cache), weights_only=True); blob["artifact"] = None; buf = io.BytesIO(); torch.save(blob, buf)` -- is repeated across ~10 tests to force the inlined (no-cache) load path. A small helper (e.g. `_strip_artifact(cache) -> bytes`) would DRY it and make the "inlined path" intent explicit at each call site.

File: torch/_precompile.py (hunk 3 above)
Comment (bobrenjc93) [reply torch/_precompile.py:1086]: already done here https://github.com/pytorch/pytorch/pull/190226/changes

**Comment by bobrenjc93** [comment]
@aorenste this PR is updated and has flaky tests so mentioning you to get on your queue!

File: test/test_precompile.py [hunk 4]
```diff
@@ -456,6 +1207,284 @@ def test_load_rejects_mismatched_code_cache_pair(self):
...
+        # inductor backend's cached AND inlined load paths. Before the fix the eager
+        # backend (no assert_size_stride backstop) silently returned a wrong-shaped tensor.
+        m = torch.nn.Linear(4, 3).eval()  # M = 3
+        x = torch.randn(5, 4)
+        code, cache = torch.compiler.precompile(
+            lambda model, t: model(t), m, x, backend=backend
+        )
+        bad = torch.nn.Linear(4, 7).eval()  # K = 7 != 3, same param names
+
+        def loaders():
```
Comment (aorenste) [inline test/test_precompile.py]: [cleanup] This `loaders()` generator is byte-for-byte identical here, in `test_param_dtype_mismatch_rejected` (line 1357), and in `test_buffer_shape_dtype_mismatch_rejected` (line 1400). A module-level helper next to `_strip_artifact` would collapse all three:

```python
def _default_and_inlined_loaders(code, cache, backend):
    yield "default", torch.compiler.precompile.load(code, cache)
    if backend == "inductor":
        yield "inlined", torch.compiler.precompile.load(code, _strip_artifact(cache))
```

then `for label, f_c in _default_and_inlined_loaders(code, cache, backend):` at each site. (The variant in `test_param_layout_specialization_rejected_inductor` at line 1437 always yields cached+inlined, so it stays as-is or takes a flag.)

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)
